File: dotnet/src/SimpleOrm.Sqlite/SqliteShadow.cs
using System.Reflection;
using System.Text;
using Microsoft.Data.Sqlite;

namespace SimpleOrm.Sqlite;

/// <summary>
/// Rebuilds the versioned schema snapshots (<c>V000N.schema.json</c>) by replaying
/// the migration history into a throwaway shadow database and introspecting the
/// touched tables after each version (ADR-0017). The full rebuild replays from
/// empty; the range form (<c>--from V000N [--to V000M]</c>) **trusts version N as
/// correct** — the baseline state is reconstructed from the committed snapshots at
/// ≤ N without verifying anything below N, versions ≤ N are baselined, and only
/// (N, M] replays and re-snapshots. Tables only; views self-reflect.
/// </summary>
public static class SqliteShadow
{
    public sealed class Result
    {
        public List<string> WrittenFiles { get; } = [];

        public List<string> Notes { get; } = [];
    }

    public static async Task<Result> RebuildSnapshotsAsync(
        Assembly assembly,
        string? migrationsNamespace,
        string migrationsDir,
        long fromVersion = 0,
        long toVersion = long.MaxValue,
        CancellationToken ct = default)
    {
        var result = new Result();
        var versions = DiscoverVersions(assembly, migrationsNamespace);
        if (versions.Count == 0)
        {
            result.Notes.Add("no migration versions found");
            return result;
        }

        var entityByRelation = MapEntities(assembly);
        var shadowFile = Path.Combine(Path.GetTempPath(), "simpleorm-shadow-" + Guid.NewGuid().ToString("n") + ".db");
        var generatedAt = DateTimeOffset.UtcNow;
        try
        {
            var db = await Db.OpenAsync(
                $"Data Source={shadowFile}", new DbOptions { Dialect = new SqliteDialect() }, ct).ConfigureAwait(false);
            await using (db.ConfigureAwait(false))
            {
                using var probe = new SqliteConnection($"Data Source={shadowFile}");
                probe.Open();

                if (fromVersion > 0)
                {
                    // Trust: rebuild the state at N from the committed snapshots, verify nothing below.
                    RestoreBaseline(probe, migrationsDir, fromVersion, result);
                    await new MigrationRunner(db, versions).BaselineAsync(fromVersion, ct).ConfigureAwait(false);
                }

                foreach (var version in versions.Where(v => v.Version > fromVersion && v.Version <= toVersion))
                {
                    var runner = new MigrationRunner(db, versions.Where(v => v.Version <= version.Version));
                    var touched = (await runner.StatusAsync(ct).ConfigureAwait(false))
                        .Where(e => e.Version == version.Version)
                        .Select(e => e.ObjectName)
                        .Distinct()
                        .ToArray();
                    await runner.MigrateAsync(ct).ConfigureAwait(false);

                    foreach (var relation in touched)
                    {
                        if (!entityByRelation.TryGetValue(relation, out var entity))
                        {
                            result.Notes.Add(
                                $"V{version.Version:0000}: '{relation}' has no mapped entity; snapshot skipped");
                            continue;
                        }

                        string? content = null;
                        if (entity.Kind == RelationKind.Table)
                        {
                            var schema = Introspect(probe, relation);
                            if (schema is not null)
                            {
                                content = SchemaSnapshot.Export(schema, version.Version, generatedAt);
                            }
                        }
                        else
                        {
                            // Views self-reflect: their history is the DDL, verbatim from sqlite_master.
                            var ddl = ViewDdl(probe, relation);
                            if (ddl is not null)
                            {
                                content = SchemaSnapshot.ExportDdl(
                                    relation, entity.KindToken, ddl, version.Version, generatedAt);
                            }
                        }

                        if (content is null)
                        {
                            continue;   // dropped at this version — its snapshot history ends here
                        }

                        var directory = Path.Combine(migrationsDir, entity.KindFolder, entity.TypeName);
                        Directory.CreateDirectory(directory);
                        var file = Path.Combine(directory, $"V{version.Version:0000}.schema.json");
                        File.WriteAllText(file, content + "\n");
                        result.WrittenFiles.Add(file);
                    }
                }
            }
        }
        finally
        {
            SqliteConnection.ClearAllPools();
            TryDelete(shadowFile);
            TryDelete(shadowFile + "-wal");
            TryDelete(shadowFile + "-shm");
        }

        return result;
    }

    // --- trusted baseline (--from) --------------------------------------------------

    private static void RestoreBaseline(SqliteConnection connection, string migrationsDir, long fromVersion, Result result)
    {
        var tableRoot = Path.Combine(migrationsDir, "Table");
        if (Directory.Exists(tableRoot))
        {
            foreach (var objectDir in Directory.GetDirectories(tableRoot))
            {
                var latest = Directory.GetFiles(objectDir, "V*.schema.json")
                    .Select(file => (File: file, Parsed: SchemaSnapshot.Parse(File.ReadAllText(file))))
                    .Where(s => s.Parsed.AsOfVersion <= fromVersion)
                    .OrderByDescending(s => s.Parsed.AsOfVersion)
                    .FirstOrDefault();
                if (latest.File is null)
                {
                    continue;   // table born after the trusted version
                }

                Execute(connection, CreateTableSql(latest.Parsed.Schema));
                foreach (var indexSql in CreateIndexSql(latest.Parsed.Schema))
                {
                    Execute(connection, indexSql);
                }

                result.Notes.Add(
                    $"baseline: {latest.Parsed.Schema.Name} restored from V{latest.Parsed.AsOfVersion:0000} snapshot (trusted)");
            }
        }
        else
        {
            result.Notes.Add($"--from V{fromVersion:0000}: no committed snapshots under {tableRoot}; starting empty");
        }

        // Views restore after tables — their DDL snapshots are executable as stored.
        foreach (var kindFolder in new[] { "View", "MaterializedView" })
        {
            var kindRoot = Path.Combine(migrationsDir, kindFolder);
            if (!Directory.Exists(kindRoot))
            {
                continue;
            }

            foreach (var objectDir in Directory.GetDirectories(kindRoot))
            {
                var latest = Directory.GetFiles(objectDir, "V*.schema.json")
                    .Select(file => SchemaSnapshot.ParseDdl(File.ReadAllText(file)))
                    .Where(s => s.AsOfVersion <= fromVersion)
                    .OrderByDescending(s => s.AsOfVersion)
                    .FirstOrDefault();
                if (latest.Ddl is null)
                {
                    continue;
                }

                Execute(connection, latest.Ddl);
                result.Notes.Add(
                    $"baseline: {latest.Object} restored from V{latest.AsOfVersion:0000} snapshot (trusted)");
            }
        }
    }

    /// <summary>The stored create statement of a view, normalized — null when absent (dropped).</summary>
    private static string? ViewDdl(SqliteConnection connection, string relation)
    {
        using var command = connection.CreateCommand();
        command.CommandText = "select sql from sqlite_master where type = 'view' and name = @name";
        command.Parameters.AddWithValue("@name", relation);
        return command.ExecuteScalar() is string sql ? SchemaSnapshot.NormalizeDdl(sql) : null;
    }

    private static string CreateTableSql(TableSchema schema)
    {
        var builder = new StringBuilder("create table if not exists ").Append(schema.Name).Append(" (");
        var first = true;
        foreach (var column in schema.Columns)
        {
            builder.Append(first ? "\n    " : ",\n    ").Append(column.Name).Append(' ');
            first = false;
            if (column.Key && column.Generated)
            {
                builder.Append("INTEGER PRIMARY KEY");   // the rowid alias spelling
                continue;
            }

            builder.Append(column.StorageType);
            if (!column.Nullable)
            {
                builder.Append(" NOT NULL");
            }
        }

        var plainKeys = schema.Columns.Where(c => c.Key && !c.Generated).ToArray();
        if (plainKeys.Length > 0 && !schema.Columns.Any(c => c.Key && c.Generated))
        {
            builder.Append(",\n    primary key (").Append(string.Join(", ", plainKeys.Select(k => k.Name))).Append(')');
        }

        return builder.Append("\n) STRICT").ToString();
    }

    private static IEnumerable<string> CreateIndexSql(TableSchema schema)
        => schema.Indexes.Select(index =>
            "create " + (index.Unique ? "unique " : string.Empty) + "index if not exists " + index.Name
            + " on " + schema.Name + " ("
            + string.Join(", ", index.Columns.Select(p => p.ColumnName + (p.Descending ? " desc" : string.Empty)))
            + ")");

    // --- introspection --------------------------------------------------------------

    /// <summary>The live shape of a table, or null when the relation is not a table (view, or dropped).</summary>
    private static TableSchema? Introspect(SqliteConnection connection, string relation)
    {
        if (!Scalar(connection, "select count(*) from sqlite_master where type = 'table' and name = @name", relation))
        {
            return null;
        }

        var columns = new List<TableSchema.Column>();
        var keyCount = 0;
        using (var command = connection.CreateCommand())
        {
            command.CommandText = "select name, type, \"notnull\", pk from pragma_table_info(@name)";
            command.Parameters.AddWithValue("@name", relation);
            using var reader = command.ExecuteReader();
            var raw = new List<(string Name, string Type, bool NotNull, long Pk)>();
            while (reader.Read())
            {
                raw.Add((reader.GetString(0), reader.GetString(1), reader.GetInt64(2) != 0, reader.GetInt64(3)));
            }

            keyCount = raw.Count(c => c.Pk > 0);
            foreach (var c in raw)
            {
                var isKey = c.Pk > 0;
                // The rowid alias: a lone INTEGER PRIMARY KEY column is database-generated.
                var generated = isKey && keyCount == 1
                    && string.Equals(c.Type, "INTEGER", StringComparison.OrdinalIgnoreCase);
                columns.Add(new TableSchema.Column(c.Name, c.Type, nullable: !c.NotNull && !isKey, isKey, generated));
            }
        }

        var indexes = new List<TableSchema.Index>();
        using (var listCommand = connection.CreateCommand())
        {
            listCommand.CommandText = "select name, \"unique\" from pragma_index_list(@name) where origin = 'c'";
            listCommand.Parameters.AddWithValue("@name", relation);
            var found = new List<(string Name, bool Unique)>();
            using (var reader = listCommand.ExecuteReader())
            {
                while (reader.Read())
                {
                    found.Add((reader.GetString(0), reader.GetInt64(1) != 0));
                }
            }

            foreach (var (name, unique) in found)
            {
                using var partsCommand = connection.CreateCommand();
                partsCommand.CommandText =
                    "select name, \"desc\" from pragma_index_xinfo(@name) where key = 1 order by seqno";
                partsCommand.Parameters.AddWithValue("@name", name);
                var parts = new List<TableSchema.Index.Part>();
                using var partsReader = partsCommand.ExecuteReader();
                while (partsReader.Read())
                {
                    parts.Add(new TableSchema.Index.Part(partsReader.GetString(0), partsReader.GetInt64(1) != 0));
                }

                indexes.Add(new TableSchema.Index(name, parts, unique));
            }
        }

        return new TableSchema(relation, columns, indexes);
    }

    // --- helpers --------------------------------------------------------------------

    private static IReadOnlyList<MigrationVersion> DiscoverVersions(Assembly assembly, string? migrationsNamespace)
        => assembly.GetTypes()
            .Where(t => !t.IsAbstract
                && typeof(MigrationVersion).IsAssignableFrom(t)
                && t.GetConstructor(Type.EmptyTypes) is not null
                && (migrationsNamespace is null
                    || (t.Namespace ?? string.Empty).StartsWith(migrationsNamespace, StringComparison.Ordinal)))
            .Select(t => (MigrationVersion)Activator.CreateInstance(t)!)
            .OrderBy(v => v.Version)
            .ToArray();

    private sealed class MappedEntity(string typeName, RelationKind kind)
    {
        public string TypeName { get; } = typeName;

        public RelationKind Kind { get; } = kind;

        public string KindFolder => Kind switch
        {
            RelationKind.Table => "Table",
            RelationKind.View => "View",
            RelationKind.MaterializedView => "MaterializedView",
            _ => "Procedure",
        };

        public string KindToken => Kind switch
        {
            RelationKind.View => "view",
            RelationKind.MaterializedView => "materialized_view",
            _ => "procedure",
        };
    }

    private static Dictionary<string, MappedEntity> MapEntities(Assembly assembly)
    {
        var loader = new EntityMapLoader();
        var byRelation = new Dictionary<string, MappedEntity>(StringComparer.OrdinalIgnoreCase);
        foreach (var type in assembly.GetExportedTypes()
            .Where(t => t is { IsClass: true, IsAbstract: false } && EntityMapLoader.HasMappingAttributes(t)))
        {
            var map = loader.Load(type);
            if (map.RelationName is not null && map.Kind != RelationKind.Statement)
            {
                byRelation[map.RelationName] = new MappedEntity(type.Name, map.Kind);
            }
        }

        return byRelation;
    }

    private static void Execute(SqliteConnection connection, string sql)
    {
        using var command = connection.CreateCommand();
        command.CommandText = sql;
        command.ExecuteNonQuery();
    }

    private static bool Scalar(SqliteConnection connection, string sql, string name)
    {
        using var command = connection.CreateCommand();
        command.CommandText = sql;
        command.Parameters.AddWithValue("@name", name);
        return (long)command.ExecuteScalar()! > 0;
    }

    private static void TryDelete(string path)
    {
        try
        {
            if (File.Exists(path))
            {
                File.Delete(path);
            }
        }
        catch (IOException)
        {
            // A straggling handle on Windows; the temp file is harmless.
        }
    }
}

File: php/src/Dialect/SqliteShadow.php
<?php

declare(strict_types=1);

namespace SimpleOrm\Dialect;

use DateTimeImmutable;
use DateTimeZone;
use PDO;
use SimpleOrm\Metadata\EntityMapLoader;
use SimpleOrm\Metadata\RelationKind;
use SimpleOrm\Migrations\MigrationRunner;
use SimpleOrm\Migrations\MigrationSet;
use SimpleOrm\Migrations\MigrationStep;
use SimpleOrm\Migrations\SchemaSnapshot;
use SimpleOrm\Migrations\SnapshotDdl;
use SimpleOrm\Migrations\SnapshotSet;
use SimpleOrm\Migrations\TableMigration;
use SimpleOrm\Migrations\TableSchema;
use SimpleOrm\Migrations\TableSchemaColumn;
use SimpleOrm\Migrations\TableSchemaIndex;
use SimpleOrm\Migrations\TableSchemaIndexPart;
use SimpleOrm\Migrations\VersionBuilder;
use SimpleOrm\Migrations\ViewMigration;

/**
 * Rebuilds the versioned schema snapshots (`V000N.schema.json`) by replaying
 * the migration history into a throwaway shadow SQLite database and
 * introspecting the touched tables/views after each version (ADR-0017),
 * mirroring `dotnet/src/SimpleOrm.Sqlite/SqliteShadow.cs`. The full rebuild
 * replays from empty; the range form (`$fromVersion`/`$toVersion`) **trusts
 * version `$fromVersion` as correct** — the baseline state is reconstructed
 * from the committed snapshots at or below it, without verifying anything
 * below it, and only `($fromVersion, $toVersion]` replays and re-snapshots.
 *
 * **PHP adaptation:** the C# reference resolves each touched relation back to
 * its mapped entity by scanning the whole assembly (`MapEntities`). PHP has no
 * assembly to scan (CODING-STANDARD §10), so this walks the other direction —
 * `TableMigration`/`ViewMigration` steps already carry `entityClass()`
 * (needed to render their own SQL), so the entity, and therefore its relation
 * name and snapshot folder, come directly from the step composing each
 * version, not from a separate assembly-wide lookup.
 */
final class SqliteShadow
{
    public static function rebuildSnapshots(
        MigrationSet $set,
        EntityMapLoader $maps,
        string $migrationsDir,
        int $fromVersion = 0,
        int $toVersion = PHP_INT_MAX,
    ): ShadowResult {
        $result = new ShadowResult();
        $versions = $set->versions();
        if ($versions === []) {
            $result->notes[] = 'no migration versions found';

            return $result;
        }

        $dialect = new SqliteDialect();
        $shadowFile = sys_get_temp_dir() . '/simpleorm-shadow-' . bin2hex(random_bytes(8)) . '.db';
        $generatedAt = new DateTimeImmutable('now', new DateTimeZone('UTC'));
        try {
            $connection = $dialect->createConnection($shadowFile);

            if ($fromVersion > 0) {
                // Trust: rebuild the state at fromVersion from the committed
                // snapshots, verify nothing below it.
                self::restoreBaseline($connection, $migrationsDir, $fromVersion, $result);
                (new MigrationRunner($connection, $dialect, $maps, MigrationSet::of(...$versions)))->baseline($fromVersion);
            }

            foreach ($versions as $version) {
                if ($version->version() <= $fromVersion || $version->version() > $toVersion) {
                    continue;
                }

                $subset = array_values(array_filter($versions, static fn ($v): bool => $v->version() <= $version->version()));
                $runner = new MigrationRunner($connection, $dialect, $maps, MigrationSet::of(...$subset));
                $runner->migrate();

                $builder = new VersionBuilder();
                $version->compose($builder);
                foreach ($builder->steps() as $step) {
                    self::snapshotStep($connection, $maps, $step, $version->version(), $migrationsDir, $generatedAt, $result);
                }
            }
        } finally {
            self::tryDelete($shadowFile);
            self::tryDelete($shadowFile . '-wal');
            self::tryDelete($shadowFile . '-shm');
            self::tryDelete($shadowFile . '-journal');
        }

        return $result;
    }

    // --- per-step snapshotting -------------------------------------------------------

    private static function snapshotStep(
        PDO $connection,
        EntityMapLoader $maps,
        MigrationStep $step,
        int $version,
        string $migrationsDir,
        DateTimeImmutable $generatedAt,
        ShadowResult $result,
    ): void {
        if ($step instanceof TableMigration) {
            $map = $maps->load($step->entityClass());
            $schema = self::introspectTable($connection, (string) $map->relationName);
            if ($schema === null) {
                return;   // dropped at this version — its snapshot history ends here
            }

            self::writeSnapshot(
                $migrationsDir,
                'Table',
                self::shortName($step->entityClass()),
                $version,
                SchemaSnapshot::export($schema, $version, $generatedAt),
                $result,
            );

            return;
        }

        if ($step instanceof ViewMigration) {
            $map = $maps->load($step->entityClass());
            $ddl = self::viewDdl($connection, (string) $map->relationName);
            if ($ddl === null) {
                return;
            }

            $materialized = $map->kind === RelationKind::MaterializedView;
            self::writeSnapshot(
                $migrationsDir,
                $materialized ? 'MaterializedView' : 'View',
                self::shortName($step->entityClass()),
                $version,
                SchemaSnapshot::exportDdl((string) $map->relationName, $materialized ? 'materialized_view' : 'view', $ddl, $version, $generatedAt),
                $result,
            );
        }
    }

    private static function writeSnapshot(
        string $migrationsDir,
        string $kindFolder,
        string $typeName,
        int $version,
        string $content,
        ShadowResult $result,
    ): void {
        $directory = "{$migrationsDir}/{$kindFolder}/{$typeName}";
        if (!is_dir($directory)) {
            mkdir($directory, 0777, true);
        }

        $file = $directory . '/' . sprintf('V%04d.schema.json', $version);
        file_put_contents($file, $content . "\n");
        $result->writtenFiles[] = $file;
    }

    // --- trusted baseline (--from) -----------------------------------------------------

    private static function restoreBaseline(PDO $connection, string $migrationsDir, int $fromVersion, ShadowResult $result): void
    {
        $tableRoot = "{$migrationsDir}/Table";
        if (is_dir($tableRoot)) {
            foreach (SnapshotSet::subdirectories($tableRoot) as $objectDir) {
                $latest = self::latestTableSnapshotAtOrBelow($objectDir, $fromVersion);
                if ($latest === null) {
                    continue;   // table born after the trusted version
                }

                [$schema, $atVersion] = $latest;
                self::execute($connection, SnapshotDdl::createTableSql($schema));
                foreach (SnapshotDdl::createAllIndexSql($schema) as $indexSql) {
                    self::execute($connection, $indexSql);
                }

                $result->notes[] = sprintf('baseline: %s restored from V%04d snapshot (trusted)', $schema->name, $atVersion);
            }
        } else {
            $result->notes[] = sprintf('--from V%04d: no committed snapshots under %s; starting empty', $fromVersion, $tableRoot);
        }

        // Views restore after tables — their DDL snapshots are executable as stored.
        foreach (['View', 'MaterializedView'] as $kindFolder) {
            $kindRoot = "{$migrationsDir}/{$kindFolder}";
            if (!is_dir($kindRoot)) {
                continue;
            }

            foreach (SnapshotSet::subdirectories($kindRoot) as $objectDir) {
                $latest = self::latestDdlSnapshotAtOrBelow($objectDir, $fromVersion);
                if ($latest === null) {
                    continue;
                }

                [$ddl, $objectName, $atVersion] = $latest;
                self::execute($connection, $ddl);
                $result->notes[] = sprintf('baseline: %s restored from V%04d snapshot (trusted)', $objectName, $atVersion);
            }
        }
    }

    /** @return array{0: TableSchema, 1: int}|null */
    private static function latestTableSnapshotAtOrBelow(string $dir, int $version): ?array
    {
        $best = null;
        $bestVersion = -1;
        foreach (glob("{$dir}/V*.schema.json") ?: [] as $file) {
            $content = file_get_contents($file);
            if ($content === false) {
                continue;
            }

            $parsed = SchemaSnapshot::parse($content);
            if ($parsed->asOfVersion <= $version && $parsed->asOfVersion > $bestVersion) {
                $bestVersion = $parsed->asOfVersion;
                $best = $parsed->schema;
            }
        }

        return $best === null ? null : [$best, $bestVersion];
    }

    /** @return array{0: string, 1: string, 2: int}|null [ddl, objectName, asOfVersion] */
    private static function latestDdlSnapshotAtOrBelow(string $dir, int $version): ?array
    {
        $best = null;
        $bestVersion = -1;
        $bestObject = null;
        foreach (glob("{$dir}/V*.schema.json") ?: [] as $file) {
            $content = file_get_contents($file);
            if ($content === false) {
                continue;
            }

            $parsed = SchemaSnapshot::parseDdl($content);
            if ($parsed->asOfVersion <= $version && $parsed->asOfVersion > $bestVersion) {
                $bestVersion = $parsed->asOfVersion;
                $best = $parsed->ddl;
                $bestObject = $parsed->object;
            }
        }

        return $best === null || $bestObject === null ? null : [$best, $bestObject, $bestVersion];
    }

    // --- introspection -------------------------------------------------------------------

    /** The live shape of a table, or null when the relation is not a table (view, or dropped). */
    private static function introspectTable(PDO $connection, string $relation): ?TableSchema
    {
        $exists = self::scalar($connection, "select count(*) from sqlite_master where type = 'table' and name = :name", $relation);
        if ($exists === 0) {
            return null;
        }

        $raw = [];
        $statement = $connection->prepare('select name, type, "notnull", pk from pragma_table_info(:name)');
        $statement->bindValue(':name', $relation, PDO::PARAM_STR);
        $statement->execute();
        while (($row = $statement->fetch(PDO::FETCH_ASSOC)) !== false) {
            $raw[] = [
                'name' => (string) $row['name'],
                'type' => (string) $row['type'],
                'notnull' => ((int) $row['notnull']) !== 0,
                'pk' => (int) $row['pk'],
            ];
        }

        $keyCount = count(array_filter($raw, static fn (array $c): bool => $c['pk'] > 0));
        $columns = [];
        foreach ($raw as $c) {
            $isKey = $c['pk'] > 0;
            // The rowid alias: a lone INTEGER PRIMARY KEY column is database-generated.
            $generated = $isKey && $keyCount === 1 && strcasecmp($c['type'], 'INTEGER') === 0;
            $columns[] = new TableSchemaColumn($c['name'], $c['type'], !$c['notnull'] && !$isKey, $isKey, $generated);
        }

        $indexes = [];
        $listStatement = $connection->prepare("select name, \"unique\" from pragma_index_list(:name) where origin = 'c'");
        $listStatement->bindValue(':name', $relation, PDO::PARAM_STR);
        $listStatement->execute();
        $found = [];
        while (($row = $listStatement->fetch(PDO::FETCH_ASSOC)) !== false) {
            $found[] = ['name' => (string) $row['name'], 'unique' => ((int) $row['unique']) !== 0];
        }

        foreach ($found as $index) {
            $partsStatement = $connection->prepare('select name, "desc" from pragma_index_xinfo(:name) where key = 1 order by seqno');
            $partsStatement->bindValue(':name', $index['name'], PDO::PARAM_STR);
            $partsStatement->execute();
            $parts = [];
            while (($row = $partsStatement->fetch(PDO::FETCH_ASSOC)) !== false) {
                $parts[] = new TableSchemaIndexPart((string) $row['name'], ((int) $row['desc']) !== 0);
            }

            $indexes[] = new TableSchemaIndex($index['name'], $parts, $index['unique']);
        }

        return new TableSchema($relation, $columns, $indexes);
    }

    /** The stored create statement of a view, normalized — null when absent (dropped). */
    private static function viewDdl(PDO $connection, string $relation): ?string
    {
        $statement = $connection->prepare("select sql from sqlite_master where type = 'view' and name = :name");
        $statement->bindValue(':name', $relation, PDO::PARAM_STR);
        $statement->execute();
        $sql = $statement->fetchColumn();

        return $sql === false ? null : SchemaSnapshot::normalizeDdl((string) $sql);
    }

    // --- helpers ---------------------------------------------------------------------------

    private static function scalar(PDO $connection, string $sql, string $name): int
    {
        $statement = $connection->prepare($sql);
        $statement->bindValue(':name', $name, PDO::PARAM_STR);
        $statement->execute();

        return (int) $statement->fetchColumn();
    }

    private static function execute(PDO $connection, string $sql): void
    {
        $connection->exec($sql);
    }

    private static function shortName(string $fqcn): string
    {
        $slash = strrpos($fqcn, '\\');

        return $slash === false ? $fqcn : substr($fqcn, $slash + 1);
    }

    /** Best-effort cleanup of the throwaway shadow database: Windows can briefly hold a delete lock after SQLite closes its handle, so a failure here is not reported (never `@` — the warning is caught and discarded explicitly, scoped to this one call). */
    private static function tryDelete(string $path): void
    {
        if (!is_file($path)) {
            return;
        }

        set_error_handler(static fn (): bool => true);
        try {
            unlink($path);
        } finally {
            restore_error_handler();
        }
    }
}
